### PR TITLE
Build My Feeds Component

### LINKED ISSUE #34: Display previous feeds on My Feeds page

### PR DESCRIPTION
**Pull request by jagjennings**
This PR adds the following:
- Replaces result page with My Feeds page, meaning once a feed is created, the user is redirected to "My Feeds" with the newly created feed highlighted
- Displays all previous feeds made by the user currently logged in
- Allows users to delete any of their previous feeds from datastore
- My Feeds page before new feed creation: https://screenshot.googleplex.com/ipv7zyQUjfR.png
- My Feeds page after new feed creation: https://screenshot.googleplex.com/JWUDXontXCM.png

### CODE REVIEW

File: launchpod/angular-launchpod/src/app/app.component.ts [hunk 1]
```diff
@@ -31,9 +31,12 @@ export class AppComponent implements OnInit {
       .subscribe((user) => {
         if (user.isLoggedIn) {
           document.getElementById("login-container").innerHTML = user.message;
+          this.formHandlerService.sendMyFeeds(user.feeds);
+          console.log(user.feeds);
```
Comment (bramanudom) [inline launchpod/angular-launchpod/src/app/app.component.ts:35]: Please remove :P

Comment (bramanudom) [inline launchpod/angular-launchpod/src/app/app.component.ts]: Please remove :P

File: launchpod/angular-launchpod/src/app/app.component.ts [hunk 2]
```diff
@@ -31,9 +31,12 @@ export class AppComponent implements OnInit {
...
         } else {
           this.loginLink = user.message;
           this.formHandlerService.sendLoginLink(user.message);
+          
           const dialogConfig = new MatDialogConfig();
```
Comment (bramanudom) [inline launchpod/angular-launchpod/src/app/app.component.ts]: nit: remove this line and you can group all the dialogconfig lines together.

File: launchpod/angular-launchpod/src/app/create-form/create-form.component.ts [hunk 3]
```diff
@@ -20,10 +49,26 @@ export class CreateFormComponent implements OnInit {
     let formData = new HttpParams();
     formData = formData.set('title', (document.getElementById("title") as HTMLInputElement).value);
     formData = formData.set('mp3Link', (document.getElementById("mp3Link") as HTMLInputElement).value);
+    formData = formData.set('name', (document.getElementById("name") as HTMLInputElement).value);
+    formData = formData.set('category', this.selected);
+
+    // this.formHandlerService.postFormData(formData)
+    //   .subscribe((response) => {
+    //     this.formHandlerService.sendFeedValue(response);
+    //   });
```
Comment (bramanudom) [inline launchpod/angular-launchpod/src/app/create-form/create-form.component.ts]: Please remove!

File: launchpod/angular-launchpod/src/app/create-form/create-form.component.ts [hunk 4]
```diff
@@ -20,10 +49,26 @@ export class CreateFormComponent implements OnInit {
...
+    .subscribe((response) => {
+      this.formHandlerService.sendMyFeeds(response);
+      this.formHandlerService.sendNewFeed();
+    });
+
+    // this.formHandlerService.getLoginData()
+    //   .subscribe((user) => {
+    //     this.formHandlerService.sendMyFeeds(user.feeds);
+    //     console.log(user.feeds);
+    //   });
```
Comment (bramanudom) [inline launchpod/angular-launchpod/src/app/create-form/create-form.component.ts]: This too!

File: launchpod/angular-launchpod/src/app/form-handler.service.ts [hunk 5]
```diff
@@ -32,11 +38,27 @@ export class FormHandlerService {
...
+  sendMyFeeds(feeds) {
+    console.log("My feeds service: " + feeds);
+    this.myFeedsSubject.next(feeds);
+  }
+
+  /**
+   * Update the newFeed boolean when a new feed is added.
+   */
+  sendNewFeed() {
+    console.log("Send New Feed Service called");
```
Comment (bramanudom) [inline launchpod/angular-launchpod/src/app/form-handler.service.ts]: Please remove

File: launchpod/angular-launchpod/src/app/form-handler.service.ts [hunk 6]
```diff
@@ -16,6 +16,12 @@ export class FormHandlerService {
   private loginLinkSubject = new BehaviorSubject<string>("Loading...");
   loginLink = this.loginLinkSubject.asObservable();
 
+  private myFeedsSubject = new BehaviorSubject<Array<any>>([]);
```
Comment (bramanudom) [inline launchpod/angular-launchpod/src/app/form-handler.service.ts]: Nit: Sorry I didn't catch this earlier, but should behavior subjects be marked as readonly? readonly is basically like declaring const

File: launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.css [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  display: inline-block;
+  height: 80vh;
+  text-align: center;
+  width: 80vw;
+  padding: 10px;
+  left: 50%;
+  position: absolute;
+  top: 50%;
+  transform: translate(-50%, -50%);
+  -ms-transform: translate(-50%, -50%);
```
Comment (bramanudom) [inline launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.css]: Please remove border attribute and keep this attributes sorted.

File: launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.ts [hunk 8]
```diff
@@ -7,8 +18,43 @@ import { Component, OnInit } from '@angular/core';
...
 
-  constructor() {}
+  hasNewFeed: boolean;
+  myFeeds: Feed[];
+
+  constructor(private formHandlerService: FormHandlerService, public snackBar: MatSnackBar) {}
+
+  ngOnInit(): void {
+    this.formHandlerService.myFeeds.subscribe((feeds) => {
+      console.log("Feeds: " + feeds);
```
Comment (bramanudom) [inline launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.ts]: Remove console.log here and below

File: launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.ts [hunk 9]
```diff
@@ -7,8 +18,43 @@ import { Component, OnInit } from '@angular/core';
...
+    });
+    
+    
+  }
+
+  // Send the key for the feed the user wants to delete to the backend.
+  public deleteFeed(key) {
+    let formData = new HttpParams();
+    formData = formData.set('keyId', key);
+    console.log("Key: " + key);
```
Comment (bramanudom) [inline launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.ts]: Same here & below

File: launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.ts [hunk 10]
```diff
@@ -7,8 +18,43 @@ import { Component, OnInit } from '@angular/core';
...
+        console.log("Feeds after delete: " + feeds);
+        this.formHandlerService.sendMyFeeds(feeds);
+        this.hasNewFeed = false;
+        this.myFeeds = feeds;
+      });
+    
+  }
 
+  openSnackBar() {
+    this.snackBar.open('Copied URL to clipboard', '', {
```
Comment (bramanudom) [inline launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.ts:54]: Nit: what is the second parameter?

Maybe try inline commenting the parameters

> Comments when calling a function
If needed, document parameters at call sites inline using block comments. Also consider named parameters using object literals and destructuring. The exact formatting and placement of the comment is not prescribed.

```
// Inline block comments for parameters that'd be hard to understand:
new Percolator().brew(/* amountLitres= */ 5);
// Also consider using named arguments and destructuring parameters (in brew's declaration):
new Percolator().brew({amountLitres: 5});
```

File: launchpod/angular-launchpod/src/app/form-handler.service.ts [hunk 11]
```diff
@@ -32,11 +38,27 @@ export class FormHandlerService {
...
+   */
+  sendMyFeeds(feeds) {
+    console.log("My feeds service: " + feeds);
+    this.myFeedsSubject.next(feeds);
+  }
+
+  /**
+   * Update the newFeed boolean when a new feed is added.
+   */
+  sendNewFeed() {
```
Comment (bramanudom) [inline launchpod/angular-launchpod/src/app/form-handler.service.ts]: Nit: I know that you are stylistically keeping all the methods send_ but the name of this is confusing. It sounds like it's supposed to do what sendMyFeeds is doing but singular, so when I saw it used in create-form it took me a second to understand what was happening. 

Consider changing?

File: launchpod/angular-launchpod/backend/src/main/java/com/google/launchpod/servlets/LoginServlet.java [hunk 12]
```diff
@@ -37,18 +53,86 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+        long timestamp = (long) entity.getProperty(LoginStatus.TIMESTAMP_KEY);
+        Date date = new Date(timestamp);
+        SimpleDateFormat dateFormat = new SimpleDateFormat("MM/dd/yyyy  HH:mm:ss Z", Locale.getDefault());
+        String postTime = dateFormat.format(date);
+        Key key = entity.getKey();
+        
+        String urlID = KeyFactory.keyToString(entity.getKey()); // the key string associated with the entity, not the numeric ID.
+        String rssLink = BASE_URL + urlID;
+
+          userFeeds.add(new UserFeed(title, name, rssLink, description, email, postTime, urlID));
```
Comment (bramanudom) [inline launchpod/angular-launchpod/backend/src/main/java/com/google/launchpod/servlets/LoginServlet.java]: Nit: match indentation

**Review by bramanudom — COMMENTED** [review]
Reviewing TS and css files only.

File: launchpod/angular-launchpod/backend/src/main/java/com/google/launchpod/servlets/FormHandlerServlet.java [hunk 13]
```diff
@@ -72,10 +107,39 @@ public void doPost(HttpServletRequest req, HttpServletResponse res) throws Illeg
     datastore.put(userFeedEntity);
 
     // return accessible link to user
-    String urlID = KeyFactory.keyToString(userFeedEntity.getKey()); // the key string associated with the entity, not the numeric ID.
-    String rssLink = BASE_URL + urlID;
-    res.setContentType("text/html");
-    res.getWriter().print(rssLink);
+    // String urlID = KeyFactory.keyToString(userFeedEntity.getKey()); // the key string associated with the entity, not the numeric ID.
```
Comment (SuperAmin) [inline launchpod/angular-launchpod/backend/src/main/java/com/google/launchpod/servlets/FormHandlerServlet.java]: remove commented code

File: launchpod/angular-launchpod/backend/src/test/java/com/google/launchpod/FormHandlerServletTest.java [hunk 14]
```diff
@@ -167,21 +195,36 @@ public void doPost_ReturnsCorrectUrl() throws IOException {
...
-    String testXmlString = RSS.toXmlString(TEST_RSS_FEED);
-    assertEquals(testXmlString, desiredEntity.getProperty(XML_STRING).toString());
-
-    String id = KeyFactory.keyToString(desiredEntity.getKey());
-    String rssLink = BASE_URL + id;
-
-    verify(response, times(1)).setContentType("text/html");
-    assertEquals(0, rssLink.compareTo(stringWriter.toString()));
+    Query query =
+        new Query(USER_FEED).addSort(TIMESTAMP, SortDirection.DESCENDING);
```
Comment (SuperAmin) [inline launchpod/angular-launchpod/backend/src/test/java/com/google/launchpod/FormHandlerServletTest.java]: you could also filter on user in this query too right?

**Review by SuperAmin — DISMISSED** [review]

File: launchpod/angular-launchpod/backend/src/test/java/com/google/launchpod/FormHandlerServletTest.java (hunk 14 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/backend/src/test/java/com/google/launchpod/FormHandlerServletTest.java]: done.

File: launchpod/angular-launchpod/src/app/app.component.ts (hunk 1 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/src/app/app.component.ts]: done.

File: launchpod/angular-launchpod/src/app/app.component.ts (hunk 2 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/src/app/app.component.ts]: done.

File: launchpod/angular-launchpod/src/app/create-form/create-form.component.ts (hunk 3 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/src/app/create-form/create-form.component.ts]: done.

File: launchpod/angular-launchpod/src/app/create-form/create-form.component.ts (hunk 4 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/src/app/create-form/create-form.component.ts]: done.

File: launchpod/angular-launchpod/src/app/form-handler.service.ts (hunk 5 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/src/app/form-handler.service.ts]: done.

File: launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.css (hunk 7 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.css]: done.

File: launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.ts (hunk 9 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.ts]: done.

File: launchpod/angular-launchpod/backend/src/main/java/com/google/launchpod/servlets/FormHandlerServlet.java (hunk 13 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/backend/src/main/java/com/google/launchpod/servlets/FormHandlerServlet.java]: done.

File: launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.ts (hunk 8 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.ts]: done.

File: launchpod/angular-launchpod/backend/src/main/java/com/google/launchpod/servlets/LoginServlet.java (hunk 12 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/backend/src/main/java/com/google/launchpod/servlets/LoginServlet.java]: done.

File: launchpod/angular-launchpod/src/app/form-handler.service.ts (hunk 11 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/src/app/form-handler.service.ts]: done.

File: launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.ts (hunk 10 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/src/app/my-feeds-page/my-feeds-page.component.ts:54]: done.

File: launchpod/angular-launchpod/src/app/form-handler.service.ts (hunk 6 above)
Comment (jagjennings) [reply launchpod/angular-launchpod/src/app/form-handler.service.ts]: done.

**Review by SuperAmin — APPROVED** [review]

**Review by kassenq — APPROVED** [review]

**Pull request merged**